File: src/main/java/net/atos/client/zgw/shared/util/AuditWijzigingJsonbDeserializer.java
/*
 * SPDX-FileCopyrightText: 2021 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.client.zgw.shared.util;

import static net.atos.client.zgw.shared.util.JsonbUtil.JSONB;

import java.lang.reflect.Type;

import jakarta.json.JsonObject;
import jakarta.json.bind.serializer.DeserializationContext;
import jakarta.json.bind.serializer.JsonbDeserializer;
import jakarta.json.stream.JsonParser;

import net.atos.client.zgw.shared.model.ObjectType;
import net.atos.client.zgw.shared.model.audit.AuditWijziging;

public class AuditWijzigingJsonbDeserializer implements JsonbDeserializer<AuditWijziging<?>> {

    @Override
    public AuditWijziging<?> deserialize(final JsonParser parser, final DeserializationContext ctx, final Type rtType) {
        final JsonObject wijzigingObject = parser.getObject();
        final JsonObject waardeObject;
        if (!wijzigingObject.isNull("oud")) {
            waardeObject = wijzigingObject.get("oud").asJsonObject();
        } else if (!wijzigingObject.isNull("nieuw")) {
            waardeObject = wijzigingObject.get("nieuw").asJsonObject();
        } else {
            return null;
        }

        final ObjectType type = ObjectType.getObjectType(waardeObject.getJsonString("url").getString());
        return JSONB.fromJson(wijzigingObject.toString(), type.getAuditClass());
    }
}

File: src/main/kotlin/nl/info/client/zgw/shared/util/AuditWijzigingJsonbDeserializer.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */

package nl.info.client.zgw.shared.util

import jakarta.json.JsonObject
import jakarta.json.bind.serializer.DeserializationContext
import jakarta.json.bind.serializer.JsonbDeserializer
import jakarta.json.stream.JsonParser
import net.atos.client.zgw.shared.model.ObjectType
import net.atos.client.zgw.shared.util.JsonbUtil.JSONB
import nl.info.client.zgw.shared.model.audit.AuditWijziging
import java.lang.reflect.Type

class AuditWijzigingJsonbDeserializer : JsonbDeserializer<AuditWijziging<*>> {
    override fun deserialize(parser: JsonParser, ctx: DeserializationContext, rtType: Type): AuditWijziging<*>? {
        // Old audit trail records (e.g. inbox/detached documents) may have wijzigingen as "" instead of {}
        val wijzigingenObject = parser.value as? JsonObject ?: return null
        val waardeObject = when {
            !wijzigingenObject.isNull("oud") -> wijzigingenObject.getJsonObject("oud")
            !wijzigingenObject.isNull("nieuw") -> wijzigingenObject.getJsonObject("nieuw")
            else -> null
        }
        return waardeObject?.let {
            JSONB.fromJson(
                wijzigingenObject.toString(),
                ObjectType.getObjectType(it.getJsonString("url").string).auditClass
            )
        }
    }
}
